resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyProviderFactory.java

```diff
@@ -2010,15 +2010,17 @@ protected void processProviderContracts(Class provider, Integer priorityOverride
       }
       if (isA(provider, Feature.class, contracts))
       {
+         ConstrainedTo constrainedTo = (ConstrainedTo) provider.getAnnotation(ConstrainedTo.class);
          int priority = getPriority(priorityOverride, contracts, Feature.class, provider);
          Feature feature = injectedInstance((Class<? extends Feature>) provider);
-         if (feature.configure(new FeatureContextDelegate(this)))
-         {
-            enabledFeatures.add(feature);
+         if (constrainedTo == null || constrainedTo.value() == getRuntimeType()) {
+            if (feature.configure(new FeatureContextDelegate(this)))
+            {
+               enabledFeatures.add(feature);
+               featureClasses.add(provider);

```

Comment:
It is the same for featureInstances. Since there is a lack of docs, I assume that purpose of this method is to get all defined enabled feature instances/classes (was not able to find any class which calls this method though). But let me put it outside the if block, so that the change is small.